keps/sig-storage/20200120-generic-data-populators.md

```diff
@@ -0,0 +1,309 @@
+
+Since the original design of the `DataSource` API field, we have been aware of the desire to
+populate volumes with data from other sources, both CSI-specific and generic (compatible with
+any CSI plugin). For new CSI-specific data sources, the path forward is clear, but for other
+sources of data, which I call "Generic Data Populators" we don't have a mechanism. The main
+problem is that current API validation logic uses a white list of object types, driven by
+feature gates for each object type. This approach won't scale to generic populators, which
+will by their nature be too numerous and varied.
+
+This proposal recommends that we relax validation on the `DataSource` field to allow
+arbitrary object types to be data sources, and leave the implementation to controllers. It
+includes example implementations to show why this is a safe thing to do.
+
+## Motivation
+
+### Goals
+
+- Enable users to create pre-populated volumes in a manner consistent with current practice
+- Enable developers to innovate with new an interesting data sources
+- Avoid changing existing workflows and breaking existing tools with new behavior
+
+### Non-Goals
+
+- In the past it was suggested that we might implement volume taints/tolerations as
+  a mechanism to facilitate the populator pod attaching to the volume before an
+  end-user pod could do so. This proposal doesn't require that.
+- In the past it was suggested that Kubernetes might implement an end-user
+  accessible interface for rebinding PVC and PVs -- in particular for swapping
+  the bindings between a pair of bound PVCs. This proposal instead relies on
+  direct modification of the PV to do rebinding. We do not think end-users
+  need a way to rebind PVCs, because it's too hard to do safely in the general
+  case.
+- We do not propose any new object types. Specifically, no special "populator"
+  object type. Populators should function based on their own object types.
+
+## Proposal
+
+The validation for `DataSource` field should be removed entirely, in the fullness of time.
+
+Short term, we should add a new alpha feature gate, which allows arbitrary objects to be
+specified for the `DataSource` field, with the intention to eventually make this behavior
+the standard. 
+
+Populators will work by responding to PVC objects with a data source they understand,
+and producing a PV with the expected data, such that ordinary Kubernetes workflows are
+not disrupted. In particular the PVC should be attachable to the end user's pod the
+moment it is bound, similar to a PVC created from currently supported data sources.
+
+### User Stories
+
+There are a long list of possible use cases around generic data populators, and I won't
+try to detail them all here. I will detail a few that illustrate the challenges faced by
+users and developers, but it's important to see these as a few examples among many.
+
+#### VM Images
+
+One use case was relevant to KubeVirt, where the Kubernetes `Pod` object is actually a
```

Comment:
Add a link to KubeVirt project.

Reply:
I'm happy to add a link, but not sure which link is best. To the main project page? To the github page?